### PR TITLE
Machine-readable `utxo_too_small` error.

### PR DESCRIPTION
**Pull request by jonathanknowles**
### Issue Number

ADP-2299

### 👤 User Story

> When using the `cardano-wallet` REST API to construct a transaction, I want to know the minimum quantity of lovelace the ledger requires for each output in the transaction.

### 📔 Release Note

> When using the `cardano-wallet` REST API to construct a transaction, if any user-specified output has a quantity of lovelace that is greater than zero but less than the minimum quantity required by the ledger, `cardano-wallet` will now return a machine-readable JSON error object with the minimum required quantity encoded as a separate field. (Previously, the minimum required quantity was embedded within an error string.)

### ℹ️ Details

This PR updates the [`UtxoTooSmall`](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/api/http/Cardano/Wallet/Api/Types.hs#L1742) error and its JSON encoding, adding an [`info`](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/specifications/api/swagger.yaml#L4905) object with the following fields:
- `tx_output_index`
   the 0-based index of the output in the `transactionOutputs` array.
- `tx_output_lovelace_specified`
   the amount of lovelace originally specified by the user.
- `tx_output_lovelace_required_minimum`
   the minimum amount of lovelace required by the ledger for this output.
   
Example of a `UtxoTooSmall` error serialized to JSON:
```json
{ "code": "utxo_too_small"
, "message": "..."
, "info":
    { "tx_output_index": 0
    , "tx_output_lovelace_specified":
        { "quantity": 1000000
        , "unit": "lovelace"
        }
    , "tx_output_lovelace_required_minimum":
        { "quantity": 1234567
        , "unit": "lovelace"
        }
    }
}
```

### 🏁 Quality Assurance

- There is now an explicit [`ApiError`](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/api/http/Cardano/Wallet/Api/Types.hs#L1641) type with a [JSON encoding round-trip test](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/test/unit/Cardano/Wallet/Api/TypesSpec.hs#L587). Previously, there was no explicit type for API errors, and our test suite did not explicitly test round-trip serialization for API errors.

- There is now a [golden JSON file](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/test/data/Cardano/Wallet/Api/ApiError.json) for the `ApiError` type that demonstrates the [extended format](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/test/data/Cardano/Wallet/Api/ApiError.json#L171) for API errors with `info` objects.

- The integration test library now has a function [`decodeErrorInfo`](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/integration/src/Test/Integration/Framework/DSL.hs#L516) that can decode the `ApiErrorInfo` associated with an API error.

- The integration test [`TRANS_MIN_UTXO_01`](https://github.com/input-output-hk/cardano-wallet/blob/9815701ed162c1ae0e02770c84c6d4a395c8b41f/lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs#L203) has been revised to demonstrate that it is possible to decode a `UtxoTooSmall` error produced as the result of an API call, and that the required minimum quantity indicated by the error can be used to successfully construct a transaction.

### CODE REVIEW

**Comment by jonathanknowles** [comment]
bors try

**Comment by jonathanknowles** [comment]
bors try

File: lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 1]
```diff
@@ -198,102 +200,66 @@ spec :: forall n.
...
-                            "unit": "lovelace"
-                        }
-                    }],
-                    "passphrase": #{fixturePassphrase}
-                }|]
+        sourceWallet <- fixtureWallet ctx
+        targetWallet <- emptyWallet ctx
+        targetAddresses <- listAddresses @n ctx targetWallet
+        let targetAddress = (targetAddresses !! 1) ^. #id
+        let endpoint = Link.createTransactionOld @'Shelley sourceWallet
```
Comment (piotr-iohk) [inline lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:216]: We test [postTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction), but the change concerns more endpoints actually, like:
 - [postTransactionFee](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransactionFee)
 - [postTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction)
 - [postByronTransactionFee](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postByronTransactionFee)
 - [postByronTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postByronTransaction)
 - [constructTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/constructTransaction)
 - [constructSharedTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/constructSharedTransaction)

Do you think it makes sense to expand this integration test or add versions of it to sub-suites for Byron, Shelley, Shared wallets for aforementioned eps?

Comment (jonathanknowles) [reply lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:216]: > We test [postTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction), but the change concerns more endpoints actually, like:
> 
>     * [postTransactionFee](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransactionFee)
> 
>     * [postTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransaction)
> 
>     * [postByronTransactionFee](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postByronTransactionFee)
> 
>     * [postByronTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postByronTransaction)
> 
>     * [constructTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/constructTransaction)
> 
>     * [constructSharedTransaction](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/constructSharedTransaction)
> 
> 
> Do you think it makes sense to expand this integration test or add versions of it to sub-suites for Byron, Shelley, Shared wallets for aforementioned eps?

Definitely! Though would you be open to us tackling these as separate tasks and/or subsequent PRs? There are quite a few integration tests with old/out-of-date/brittle methods of computing minimum UTxO values, and I'd hope to tackle those too.

I hope to create a JIRA technical debt ticket for this first thing next week. Hopefully we could assign it by next sprint at the latest. Would that work for you?

**Review by piotr-iohk — APPROVED** [review]
LGTM. Added respective checks in e2e suite.

**Comment by jonathanknowles** [comment]
> Added respective checks in e2e suite.

Thank-you @piotr-iohk!

**Comment by jonathanknowles** [comment]
bors cancel

**Pull request merged**